### DISCUSSION TITLE
0.6.0-beta 2

### DISCUSSION THREAD

**Discussion by mdelobelle** [mdelobelle/metadatamenu#421] [Announcements]
Reminder for new features in this beta: https://github.com/mdelobelle/metadatamenu/discussions/420#discussion-5813183

Fixes released in this version:

- Add ObjectList <missing> with Add field at section,
- Add ObjectList <missing> in Note field modal
- Display the fields by hierarchy in fileclass fields view and preset fields settings
- exception management for updateLookups
- update formula mixing 0 and undefined triggering "changed" status
- command "update all formulas and lookups"
- mapWithTag with tags defined in frontmatter with `tags: ....`
- legacyGenuineKey throwing error with null dvFile
- manage both tags in frontmatter (tags: ...) and in file (#....)

<hr /><em>This discussion was created from the release <a href='https://github.com/mdelobelle/metadatamenu/releases/tag/0.6.0-beta-2'>0.6.0-beta 2</a>.</em>

**Comment by arminta7** [comment]
error trying to update

<img width="340" alt="Screenshot 2023-11-05 at 14 29 58" src="https://github.com/mdelobelle/metadatamenu/assets/7927356/6509c26a-1205-4d3b-856d-35c804e85888">

**Reply by arminta7** [reply]
Nevermind, my mistake!!

**Comment by arminta7** [comment]
Bug: Can't delete auto-update relation items. Sometimes they repopulate, sometimes fields get stuck with previous values that have been deleted. Sometimes I can delete but my related field in the other file doesn't update to reflect the delete.

**Reply by mdelobelle** [reply]
is it occuring with lookups and formulas or just lookups or just formulas?

**Reply by arminta7** [reply]
At this point I've only tried this specifically with lookups.

**Comment by arminta7** [comment]
Error that popped up:

<img width="341" alt="Screenshot 2023-11-05 at 15 13 08" src="https://github.com/mdelobelle/metadatamenu/assets/7927356/5aedca26-67aa-4d39-9847-7fa73204b93c">

**Reply by mdelobelle** [reply]
I'll fix this

**Reply by mdelobelle** [reply]
I've updated the package. Could you grab the main.js [here](https://github.com/mdelobelle/metadatamenu/releases/tag/0.6.0-beta-2) and replace it in <your_vault>/.obsidian/plugins/metadata-menu and see if it fixes the bug?

**Reply by arminta7** [reply]
Updated the file, there's still weirdness. Tried to share a video of what's happening, but it's too big. Let me know if this works: https://www.dropbox.com/scl/fi/tsdbonsu00uk1ng3udgf9/Screen-Recording-2023-11-05-at-15.41.52.mov?rlkey=tdalj6enfdnrrbmqgjzzla7vx&dl=0

**Reply by mdelobelle** [reply]
I can open the video thanks.
can you paste the definition of your "goal" and "project" fileClasses? 
I can't figure out which one is a lookup and which one is a multi file select field between the two fields "Projects" and "Goals"

**Reply by arminta7** [reply]
Goal file:

```---
fields:
  - id: 4L9M5U
    name: Projects
    options:
      autoUpdate: true
      outputType: LinksList
      builtinSummarizingFunction: Count
      customListFunction: page.file.name
      customSummarizingFunction: return pages.length
      dvQueryString: dv.pages('#project')
      targetFieldName: Goals
    type: Lookup
    path: ""
version: "2.2"
limit: 100
mapWithTag: true
icon: target
tagNames: 
filesPaths: 
bookmarksGroups: 
excludes: 
extends: 
---

Projects:: {"type":"Lookup","options":{"autoUpdate":true,"outputType":"LinksList","builtinSummarizingFunction":"Count","customListFunction":"page.file.name","customSummarizingFunction":"return pages.length","dvQueryString":"dv.pages('#project')","targetFieldName":"Goals"}}
```

Project file:

```
---
mapWithTag: true
fields:
  - id: zVY0Lo
    name: Goals
    options:
      autoUpdate: true
      outputType: LinksList
      builtinSummarizingFunction: Count
      customListFunction: page.file.name
      customSummarizingFunction: return pages.length
      dvQueryString: dv.pages('#goal')
      targetFieldName: Projects
    type: Lookup
    path: ""
version: "2.2"
limit: 100
icon: folder
tagNames: 
filesPaths: 
bookmarksGroups: 
excludes: 
extends: 
---

Goals:: {"type":"Lookup","options":{"autoUpdate":true,"outputType":"LinksList","builtinSummarizingFunction":"Count","customListFunction":"page.file.name","customSummarizingFunction":"return pages.length","dvQueryString":"dv.pages('#goal')","targetFieldName":"Projects"}}
Status:: {"type":"Input","options":{}}
Priority:: {"type":"Select","options":{"valuesList":{"1":"P0","2":"P1","3":"P2","4":"P3"},"sourceType":"ValuesList","valuesListNotePath":"","valuesFromDVQuery":""}}
Tasks:: {"type":"Lookup","options":{"autoUpdate":false,"outputType":"LinksList","builtinSummarizingFunction":"Count","customListFunction":"page.file.name","customSummarizingFunction":"return pages.length","dvQueryString":"dv.pages('#task')","targetFieldName":"Project"}}
Open Tasks:: {"type":"Lookup","options":{"autoUpdate":false,"outputType":"LinksList","builtinSummarizingFunction":"CountAll","customListFunction":"page.file.name","customSummarizingFunction":"return pages.length","targetFieldName":"Project","dvQueryString":"dv.pages(\"#task\").where(p => p.completed == false)"}}
Number of Open Tasks:: {"type":"Formula","options":{"autoUpdate":false,"formula":"current[\"Open Tasks\"].length"}}
Task Progress:: {"type":"Formula","options":{"autoUpdate":false,"formula":"Math.round(((current[\"Tasks\"].length - current[\"Open Tasks\"].length)/current[\"Tasks\"].length) * 100) + \"%\""}}
Dreams:: {"type":"Formula","options":{"autoUpdate":false,"formula":"dv.page(current.Goals)['Dreams'];"}}
Life Areas:: {"type":"Formula","options":{"autoUpdate":false,"formula":"dv.page(current.Dreams)['Life Areas'];"}}
Goal Progress:: {"type":"Formula","options":{"autoUpdate":false,"formula":"dv.page(current.Goals)['Progress'];"}}
Active Child Count:: {"type":"Formula","options":{"autoUpdate":false,"formula":"current[\"Open Tasks\"].length"}}```

**Reply by mdelobelle** [reply]
I think I understand what's happening.
A lookup field can't reference another lookup field as its "targetFieldName"
In your example:
The "Projects" field in goal files (goal1, goal2) is looking in pages with the "#project" tag (project1, project2) and returning the value in the Goals field of those project files
But the "Goals" field which is also a lookup field is doing the same thing: it's looking for the value contained in "Projects" field of "#goal" files.
I've put a mecanism to avoid infinite loops (each file updating a field that trigger a recalculation of the target value would lead to infinite loop in this case) - that's why values are updating in a weird way
Moreover lookup fields aren't meant to be "manually" modified

You would have to choose one of the two class (either #goal or #project) to be manually modified with a "multi file" field, and have the other one define the Lookup field.

For example: don't change your #goal definition , but in #project change the Goals field to a Multi File (with the same dv.query)
When you will update the value in a "#project" file, it will update the "Projects" fields in the "#goals" files. But you shouldn't modify the Projects field directly in your "#goal" files.

**Reply by arminta7** [reply]
Hmm... that would be a huge change in workflow and honestly have me probably looking for another app to use. I'm trying to mimic the bidirectional relations that apps like Notion, Fibery, Coda have. I never know which side I'm going to be updating from and I have to keep everything updated no matter where the data is entered. This is the main functionality that metadata menu has enabled for me and because MM has allowed me to do this is one of the only reasons I can use Obsidian! Will this ever be supported by metadata menu natively?

**Reply by mdelobelle** [reply]
I think this is feasible. It wasn't meant to work this way by design (I'm surprised that it doesn't cause infinite loops with the current version 😅 ) so I have to make sure that I understand your use case correctly before trying to enable it as a legit use case.

If I understand correctly, when a lookup field is manually updated, we should first recalculate other lookup fields referencing it before recalculating itself (which causes the issues that you see)

**Reply by mdelobelle** [reply]
I think this is feasible. It wasn't meant to work this way by design (I'm surprised that it doesn't cause infinite loops with the current version 😅 ) so I have to make sure that I understand your use case correctly before trying to enable it as a legit use case.

If I understand correctly, when a lookup field is manually updated, we should first recalculate other lookup fields referencing it before recalculating itself (which causes the issues that you see)

**Reply by arminta7** [reply]
First class support for this would be a huge QOL improvement for me. Let me know if I can help explain further my use case. I DMd you my file class folder.

**Comment by arminta7** [comment]
Another error:

 <img width="349" alt="Screenshot 2023-11-05 at 15 35 45" src="https://github.com/mdelobelle/metadatamenu/assets/7927356/937958c0-0af4-4edc-8628-8c5bc65b4cfc">

**Reply by mdelobelle** [reply]
I suppose this occurs when opening a "Select" or "Multi Select" field modal.
Could you paste the definition of this field (or fileClass)?

**Reply by arminta7** [reply]
I need to play around more to try and figure out which vault it was happening in. It might be my bigger one and I think you added a lot more console logging so it's harder to see errors lol. I'll come back to this if/when I see it again.

**Reply by mdelobelle** [reply]
oops. I've removed the logging and uploaded a new main.js (same place)
